Dockerfile

```diff
@@ -1,44 +1,63 @@
-FROM python:3.7.3-alpine
+FROM python:3.7-slim AS python-builder
 
-# Default to building for local development.
-# Override with `--build-arg PIP_FILE=build.txt` to build for production.
-ARG PIP_FILE=all.txt
-
-EXPOSE 8000
+ENV PIP_DISABLE_PIP_VERSION_CHECK=1
+ENV PATH="/venv/bin:$PATH"
 
 WORKDIR /app
 
-RUN addgroup -g 10001 app && \
-    adduser -D -u 10001 -G app -h /app -s /sbin/nologin app
-
-RUN apk --no-cache add \
-    build-base \
-    bash \
-    curl \
-    git \
+RUN apt-get update && \
+    apt-get -y upgrade && \
+    apt-get -y install \ 
+    libpq-dev \
     libffi-dev \
-    py-cffi \
-    postgresql \
-    postgresql-dev \
-    postgresql-client
-
-# Install Python dependencies
-COPY requirements/*.txt /tmp/requirements/
-# Switch to /tmp to install dependencies outside home dir
-WORKDIR /tmp
-# TODO: Consider a way to install only the "build.txt" deps for production.
-RUN pip install --require-hashes --no-cache-dir -r requirements/$PIP_FILE
+    python-dev \
+    build-essential
+
+# Install virtualenv
+RUN pip install virtualenv
+RUN virtualenv /venv
+
+# Install base python dependencies 
+COPY requirements/build.txt ./requirements/
+RUN pip install --require-hashes --no-cache-dir -r requirements/build.txt
+
+FROM python:3.7-slim AS base
+
+ENV PIP_DISABLE_PIP_VERSION_CHECK=1
+ENV PATH="/venv/bin:$PATH"
+
+RUN groupadd --gid 10001 app && useradd -g app --uid 10001 --shell /usr/sbin/nologin app
+
+COPY --from=python-builder /venv /venv
+
+EXPOSE 8000
 
 WORKDIR /app
 COPY . /app
-RUN chown app:app -R .
-USER app
 
+# Collect static files
 RUN DEBUG=False SECRET_KEY=foo ./manage.py collectstatic --noinput -c
 
 # Using /bin/bash as the entrypoint works around some volume mount issues on Windows
 # where volume-mounted files do not have execute bits set.
 # https://github.com/docker/compose/issues/2301#issuecomment-154450785 has additional background.
 ENTRYPOINT ["/bin/bash", "/app/bin/run"]
 
-CMD ["dev"]
+FROM base AS release
+
+COPY --from=base --chown=app:app /app/static /app/static
+
+# Set user and user permissions
+COPY --chown=app:app . . 
+USER app
+
+FROM base AS devapp
+
+COPY --from=base --chown=app:app /venv /venv
+COPY --from=base --chown=app:app /app/static /app/static 
```

Comment:
Hey @wlach,

I believe you are right about the static files, although one issue that I was running into with the virtual env is if apply permissions before you pip install the dependencies will install to app/local instead which results in things like CI not being able to find certain commands. This is also why I apply permissions at the release stage. I think a similar issue with dependencies like isort is found with mozilla's Kitsune [https://github.com/mozilla/kitsune/blob/master/Dockerfile](https://github.com/mozilla/kitsune/blob/master/Dockerfile). They also do multi stage builds using virtualenv although it is specific to python2.

Reply:
I haven't tested this but I believe since the virtualenv is in your path, any new dependencies should be installed into the virtualenv. You will want to do this after the `USER app` declaration, so they are installed with the correct permissions.

Reply:
Hey @wlach,

Thank you for the suggestions, i fixed most of the inheritance. 

I think the issue I was running into was really setting the User before any of the installs. I could not install the dependencies to the user level because some of the packages were being checked at the system level. 

In the case that you do not provide a --target I just made sure that if the user is prematurely set at the release stage then the devapp stage will first set the user to root to make sure the dev dependencies are installed at the system level before setting the user back to app. This way you can run either stage individually via docker build kit, individually with only target set, or all without providing a target.